### PR TITLE
PIM-6329: Remove family variant

### PR DESCRIPTION
**Pull request by SamirBoulil**
**Description (for Contributor and Core Developer)**
**Reviewable by commit**

Be able to remove the a family variant if, and only if, there is no product models / variant products belonging to it.

- [x] Add translations in popup
- [x] Add acls

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Ok
| Added legacy Behats               | Ok
| Added acceptance tests            | -
| Added integration tests           | Ok
| Changelog updated                 | Ok
| Review and 2 GTM                  | Ok
| Micro Demo to the PO (Story only) | Ok
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+/**
+ * Delete extension for family variants
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js]: Samir ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
```
Comment (willy-ahva) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js]: 2018 :)

File: src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     *
+     * @return bool
+     */
+    private function hasEntityWithFamilyVariant(FamilyVariantInterface $familyVariant): bool
+    {
+        return 0 !== $this->counter->belongingToFamilyVariant($familyVariant);
+    }
+
+    /**
+     * @param       $familyVariant
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php]: FamilyVariantInterface

Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php]: done

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js (hunk 2 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js]: done

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js (hunk 1 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/family-variant/form/delete.js]: done

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyVariantController.php [hunk 4]
```diff
@@ -62,7 +70,8 @@ public function __construct(
         ObjectUpdaterInterface $updater,
         ValidatorInterface $validator,
         NormalizerInterface $constraintViolationNormalizer,
-        SaverInterface $saver
+        SaverInterface $saver,
+        RemoverInterface $remover = null
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyVariantController.php]: It is on 2.2, so no need to add a `= null`.

File: src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                sprintf(
+                    'Family variant "%s", could not be removed as it is used by some entities with family variants.',
+                    $familyVariant->getCode()
+                )
+            );
+        }
+
+        $this->removeFamilyVariant($familyVariant, $options);
+
+        return $this;
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php]: Is there a real use to return the Remover?

Comment (SamirBoulil) [reply src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php]: I don't think so yet, most of removers/updaters are doing it, so I do it too out of consistency.

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyVariantController.php (hunk 4 above)
Comment (SamirBoulil) [reply src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyVariantController.php]: yep!

File: src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php (hunk 5 above)
Comment (damien-carcel) [reply src/Pim/Component/Catalog/Updater/Remover/FamilyVariantRemover.php]: OK then. But imho none of them should.

File: features/family/delete_family_variant.feature [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  Scenario: Successfully delete a family variant used by product models from the grid
+    Given the following family variants:
+      | code                 | family   | variant-axes_1 | variant-attributes_1 |
+      | empty_family_variant | clothing | color          | description          |
+    And I am on the "Clothing" family page
+    And I visit the "Variants" tab
+    When I click on the "Delete" action of the row which contains "empty_family_variant"
+    And I confirm the deletion
+    Then I should see the flash message "Family variant successfully removed"
```
[the remark this answers is not in the document]
Comment (SamirBoulil) [reply features/family/delete_family_variant.feature]: Well, the step is used over 70 times so I guess it's an issue we had with the old CI

[the remark this answers is not in the document]
Comment (ahocquard) [reply features/family/delete_family_variant.feature]: Seen IRL: Function is deprecated and do nothing (early return) 🎉

**Pull request merged**